In [147]:
# based on https://hadrienj.github.io/posts/Deep-Learning-Book-Series-2.7-Eigendecomposition/

In [148]:
import numpy as np
import matplotlib.pyplot as plt

If a transformation of a vector by a matrix gives a new vector with the exact same direction as the initial vector, this vector is called an *eigenvector* of A.


Av=λv


In [149]:
A = np.array([[5., 1.], [3., 3.]], dtype =float)
A

array([[5., 1.],
       [3., 3.]])

In [150]:
eig_a = np.linalg.eig(A)
eig_a

(array([6., 2.]), array([[ 0.70710678, -0.31622777],
        [ 0.70710678,  0.9486833 ]]))

In [151]:
V = eig_a[1]
V

array([[ 0.70710678, -0.31622777],
       [ 0.70710678,  0.9486833 ]])

In [152]:
V_inv = np.linalg.inv(V)
V_inv

array([[ 1.06066017,  0.35355339],
       [-0.79056942,  0.79056942]])

We can see that the eigenvalues are the same than the ones we used before: 6 and 2 (first array).

The eigenvectors correspond to the columns of the second array.<br><br>
This means that the eigenvector corresponding to λ=6 is: [0.70710678, 0.70710678] <br>
and the eigenvector corresponsing to λ=2 is: [-0.31622777, 0.9486833]

Eigendecomposition is given by A = V * diag(λ) * V**(-1). <br>
Let's see if it works!

In [153]:
lambdas_diag = np.diag(eig_a[0])
lambdas_diag

array([[6., 0.],
       [0., 2.]])

In [154]:
A

array([[5., 1.],
       [3., 3.]])

In [155]:
V.dot(lambdas_diag).dot(V_inv)

array([[5., 1.],
       [3., 3.]])

In [156]:
np.isclose(V.dot(lambdas_diag).dot(V_inv), A)

array([[ True,  True],
       [ True,  True]])

*boom.*